File: apis_c#/AthleteApi/AthleteApi/Services/TournamentParticipationService.cs
using System.Data;
using Microsoft.Data.SqlClient;
using AthleteApi.Models;

namespace AthleteApi.Services
{
    // Implementación del servicio de participación en torneos que maneja las operaciones CRUD para las participaciones
    public class TournamentParticipationService : ITournamentParticipationService
    {
        private readonly string _connectionString;

        // Constructor que inicializa la cadena de conexión a la base de datos
        public TournamentParticipationService(IConfiguration configuration)
        {
            _connectionString = configuration?.GetConnectionString("DefaultConnection") ?? throw new ArgumentNullException(nameof(configuration), "Configuration cannot be null");
        }

        // Método para agregar un nuevo participante en un torneo
        public async Task AddParticipant(TournamentParticipation participation)
        {
            // Crea una nueva conexión SQL usando la cadena de conexión
            using (SqlConnection conn = new SqlConnection(_connectionString))
            {
                // Crea un nuevo comando SQL para ejecutar el procedimiento almacenado 'sp_addParticipant'
                using (SqlCommand cmd = new SqlCommand("sp_addParticipant", conn))
                {
                    cmd.CommandType = CommandType.StoredProcedure;

                    // Agrega los parámetros necesarios para el procedimiento almacenado
                    cmd.Parameters.AddWithValue("@AthleteId", participation.AthleteId);
                    cmd.Parameters.AddWithValue("@TournamentId", participation.TournamentId);
                    cmd.Parameters.AddWithValue("@WeightCategoryId", participation.WeightCategoryId);

                    // Abre la conexión a la base de datos
                    conn.Open();
                    // Ejecuta el comando de forma asíncrona
                    await cmd.ExecuteNonQueryAsync();
                }
            }
        }

        // Método para obtener una lista de participantes con paginación y filtro opcional por torneo y nombre del atleta
        public async Task<IEnumerable<TournamentParticipation>> GetParticipants(int? tournamentId, string? athleteName, int pageNumber, int pageSize)
        {
            var participants = new List<TournamentParticipation>();

            // Crea una nueva conexión SQL usando la cadena de conexión
            using (SqlConnection conn = new SqlConnection(_connectionString))
            {
                // Crea un nuevo comando SQL para ejecutar el procedimiento almacenado 'sp_getParticipants'
                using (SqlCommand cmd = new SqlCommand("sp_getParticipants", conn))
                {
                    cmd.CommandType = CommandType.StoredProcedure;

                    // Agrega los parámetros necesarios para el procedimiento almacenado
                    cmd.Parameters.AddWithValue("@TournamentId", tournamentId.HasValue ? (object)tournamentId.Value : DBNull.Value);
                    cmd.Parameters.AddWithValue("@AthleteName", athleteName ?? (object)DBNull.Value);
                    cmd.Parameters.AddWithValue("@PageNumber", pageNumber);
                    cmd.Parameters.AddWithValue("@PageSize", pageSize);

                    // Abre la conexión a la base de datos
                    conn.Open();
                    // Ejecuta el comando y obtiene un lector de datos de forma asíncrona
                    using (SqlDataReader reader = await cmd.ExecuteReaderAsync())
                    {
                        // Lee los datos devueltos por el lector de datos
                        while (await reader.ReadAsync())
                        {
                            var participation = new TournamentParticipation
                            {
                                Id = reader.GetInt32(0),
                                AthleteId = reader.GetInt32(1),
                                TournamentId = reader.GetInt32(2),
                                WeightCategoryId = reader.GetInt32(3)
                            };
                            // Agrega la participación a la lista de participantes
                            participants.Add(participation);
                        }
                    }
                }
            }

            // Retorna la lista de participantes
            return participants;
        }
    }
}

File: apis_py/app/services/tournament_participation_service.py
from typing import List, Optional
from fastapi import Depends
from sqlalchemy.orm import Session
from sqlalchemy import text
from app.schemas.tournament_participation import TournamentParticipationCreate, TournamentParticipationResponse
from app.database import get_db
from app.models.tournament_participation import TournamentParticipation

class TournamentParticipationService:
    """
    Servicio para manejar operaciones relacionadas con participaciones en torneos en la base de datos.
    
    Atributos:
        db (Session): Sesión de la base de datos para realizar las operaciones.
    """

    def __init__(self, db: Session = Depends(get_db)):
        """
        Inicializa el servicio con una sesión de la base de datos.
        
        Args:
            db (Session): Sesión de la base de datos inyectada automáticamente por FastAPI.
        """
        self.db = db

    def get_participants(self, tournament_id: Optional[int], athlete_name: Optional[str], page_number: int, page_size: int) -> List[TournamentParticipationResponse]:
        """
        Obtiene una lista paginada de participantes en un torneo desde la base de datos.

        Args:
            tournament_id (Optional[int]): Identificador del torneo.
            athlete_name (Optional[str]): Nombre del atleta.
            page_number (int): Número de la página a obtener.
            page_size (int): Cantidad de registros por página.

        Returns:
            List[TournamentParticipationResponse]: Lista de respuestas con la información de los participantes.
        """
        participants = self.db.execute(
            text("EXEC sp_getparticipants @TournamentId = :tournament_id, @AthleteName = :athlete_name, @PageNumber = :page_number, @PageSize = :page_size"),
            {'tournament_id': tournament_id, 'athlete_name': athlete_name, 'page_number': page_number, 'page_size': page_size}
        ).fetchall()
        return [TournamentParticipationResponse.model_validate(participant) for participant in participants]

    def add_participant(self, participation: TournamentParticipationCreate) -> TournamentParticipationResponse:
        """
        Agrega un nuevo participante en un torneo en la base de datos.

        Args:
            participation (TournamentParticipationCreate): Objeto que contiene los datos de la participación a crear.

        Returns:
            TournamentParticipationResponse: Objeto que representa la participación creada.

        Raises:
            ValueError: Si la participación ya existe en la base de datos.
        """
        self.db.execute(
            text("EXEC sp_addparticipant @AthleteId = :athlete_id, @TournamentId = :tournament_id, @WeightCategoryId = :weight_category_id"),
            {'athlete_id': participation.athlete_id, 'tournament_id': participation.tournament_id, 'weight_category_id': participation.weight_category_id}
        )
        self.db.commit()
        new_participation = self.db.query(TournamentParticipation).filter_by(athlete_id=participation.athlete_id, tournament_id=participation.tournament_id).first()
        return TournamentParticipationResponse.model_validate(new_participation)